core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -283,11 +285,30 @@ void handleNotInUse() {
 
   @Override
   public ListenableFuture<ChannelStats> getStats() {
-    SettableFuture<ChannelStats> ret = SettableFuture.create();
-    ChannelStats.Builder builder = new Channelz.ChannelStats.Builder();
+    final SettableFuture<ChannelStats> ret = SettableFuture.create();
+    final ChannelStats.Builder builder = new Channelz.ChannelStats.Builder();
     channelCallTracer.updateBuilder(builder);
     builder.setTarget(target).setState(channelStateManager.getState());
-    ret.set(builder.build());
+    // subchannels and oobchannels can only be accessed from channelExecutor
+    channelExecutor.executeLater(new Runnable() {
+      @Override
+      public void run() {
+        List<LogId> children = new ArrayList<LogId>();
+        for (InternalSubchannel internalSubchannel : subchannels) {
+          children.add(internalSubchannel.getLogId());
+        }
+        for (OobChannel oobChannel : oobChannels) {
+          children.add(oobChannel.getLogId());
+        }
+        builder.setSubchannels(children);
+        ret.set(builder.build());
+      }
+    });
+    if (!ret.isDone()) {
```

Comment:
This will always be true.  ChannelExecutor won't start until drain is called.

Reply:
I was perhaps microoptimizing for the case where we raced with a gRPC thread, then we can avoid calling drain. Since we are potentially a low pri application thread, we don't want to risk taking on important work from the queue.

Reply:
Don't optimize for that case.  In the case drain() isn't called very soon after, this may not execute.   It's like not calling flush and hoping someone else does.